lib/serialization/README.md

```diff
@@ -0,0 +1,237 @@
+The method should only act on known class names, and call super otherwise.
+
+### Example: the User class
+
+The following example cannot use the `auto_serializable` annotations
+because some of the arguments to the `User` class need special treatment:
+
+* The `name` attribute is perfectly normal, it can be serialized and deserialized
+  directly.
+
+* The `password` attribute must be "encrypted" before being serialized.
+  On deserialization it must de unencrypted.
+  Obviously, you should __never__ handle clear text password.
+
+* The `avatar` attributes is kept as a data blob in memory.
+  It could be serialized as such but it is cleaner to only
+  serialize the path ot its source on the file system.
+  The data is reloaded on deserialization.
+
+For this customization, the following code snippet implements
+two serialization services: `User::core_serialize_to` and
+`Deserializer::deserialize_class`.
+
+~~~
+# User credentials for a website
+class User
+	super Serializable
+
+	# User name
+	var name: String
+
+	# Clear text password (do not do this)
+	var password: String
+
+	# User's avatar image as data blob
+	var avatar: Image
+
+	redef fun core_serialize_to(serializer: Serializer)
+	do
+		# This is the normal serialization process
+		serializer.serialize_attribute("name", name)
+
+		# Serialized an "encrypted" version of the password
+		serializer.serialize_attribute("pass", password.rot(13))
+
+		# Do not serialize the image, only its path
+		serializer.serialize_attribute("avatar_path", avatar.path)
+	end
+end
+
+redef class Deserializer
+	redef fun deserialize_class(name)
+	do
+		if name == "User" then
+			# Deserialize normally
+			var user = deserialize_attribute("name")
+
+			# Decrypt password
+			var pass = deserialize_attribute("pass").rot(-13)
+
+			# Deserialize the path nad load the avatar from the file system
+			var avatar_path = deserialize_attribute("avatar_path")
+			var avatar = new Image(avatar_path)
+
+			return new User(user, pass, avatar)
+		end
+
+		return super
+	end
+end
+
+# An image loaded in memory
+#
+# Not really useful for this example, provided for consistency only.
+class Image
+	# Path on the filesystem for `self`
+	var path: String
+
+	# Data blob as a string of bytes (approximately a `String`)
+	var data: String = path.read_all is lazy
+end
+
+~~~
+
+See the documentation of the module `serialization::serialization` for more
+information on the services to redefine.
+
+## Serialization services
+
+The `auto_serializable` annotation and the `Serializable` class are used on
+classes specific to the business domain.
+To write (and read) instances of these classes to a persistent format
+you must use implementations of `Serializer` and `Deserializer`.
+
+The main implementations of these services are `JsonSerializer` and `JsonDeserializer`,
+from the `json_serialization` module.
+
+~~~
+import the_previous_code_in_this_README
```

Comment:
TODO: find a way to express this that can be understand with nitunit. (related to #1201)

Idea: use a `module` clause with a name in the previous code then import the name. I think it is something that by understood by the documentation reader and could be managed by nitunit.